Ta activity stats

finnergizer opened the pull request:
Current stats provided for admins and graders in the submission status column providing info on assignments completed and average number of annotations per submission that has been marked as having been 'completed' marking.

See https://slack-files.com/T0AGA2MQX-F0CPFN7QE-cabb89d8c8 for details.

Also see https://slack-files.com/T0AGA2MQX-F0DMA2NRX-5d7dc9c897 for more details on TA grade distribution

david-yz-liu commented on app/models/admin.rb:
```diff
@@ -7,4 +7,41 @@ class Admin < User
+  # for an admin, we want all the assignment groupings marked completed
+  def get_num_marked(assignment)
+    n = 0
+    assignment.groupings.each do |x|
```
I think you can use the `count` method here

david-yz-liu commented on app/models/ta.rb:
```diff
@@ -29,6 +29,28 @@ def get_num_marked(assignment)
+      num_annotations = get_num_annotations(assignment)
+      avg = num_annotations / num_marked
+    end
+    avg
```
You should round this.

david-yz-liu commented on config/locales/en.yml:
```diff
@@ -455,6 +455,7 @@ en:
       tags_list:            "Filter by Tag:"
       search_submissions:   "Search Submissions"
       how_many_marked:      "%{num_marked}/%{num_assigned} assignments have been marked"
+      average_annotations:  "%{average_annotations} annotations per marked submission"
```
Please add this to the other locale files.

finnergizer replied:
Added and translated. Thanks!

arkon commented on app/views/assignments/_list_manage.html.erb:
```diff
@@ -7,77 +7,105 @@
+                  <li><span id='class_average__<%= assignment.short_identifier %>'>
+
                     <%= t(:class_average,
                           results_average: assignment.results_average) %>
```
I think it'd be nice to have a % sign afterwards (assuming it's a percentage average). @david-yz-liu : thoughts?

david-yz-liu replied:
I agree. Also, @finnergizer can you make sure this is rounded to two decimal places?

finnergizer replied:
Added, and also checked that rounding is to two decimal places when it is calculated server-side.

david-yz-liu commented on app/controllers/main_controller.rb:
```diff
@@ -160,10 +160,21 @@ def index
     render :index, layout: 'content'
   end
 
+  def grader_summary
```
This should be a method in the `Assignment` controller

david-yz-liu commented on app/controllers/tas_controller.rb:
```diff
@@ -95,4 +95,17 @@ def upload_ta_list
     params.require(:user).permit(:user_name, :last_name, :first_name)
   end
+
+  before_action :set_assignment_ta, only: [:refresh_graph]
```
Don't make a separate method; put the two lines inside the `refresh_graph` method.

david-yz-liu commented on app/controllers/tas_controller.rb:
```diff
@@ -95,4 +95,17 @@ def upload_ta_list
+  end
+
+  def refresh_graph
+    respond_to do |format|
```
We generally haven't used `respond_to` with only one format; you can remove this.

david-yz-liu commented on app/models/admin.rb:
```diff
@@ -7,4 +7,36 @@ class Admin < User
   after_update  :maintain_repository_permissions
 
+  # for an admin, we just want all the assignment groupings submitted
+  def get_num_assigned(assignment)
```
Make this a method in the `Assignment` model.

david-yz-liu commented on app/models/admin.rb:
```diff
@@ -7,4 +7,36 @@ class Admin < User
+
+  # for an admin, we want all the assignment groupings marked completed
+  def get_num_marked(assignment)
+    n = assignment.groupings.count(marking_completed: true)
```
Don't use a temporary variable here.

david-yz-liu commented on app/models/admin.rb:
```diff
@@ -7,4 +7,36 @@ class Admin < User
+    assignment.groupings.each do |x|
+      # only grab annotations from groupings where marking is completed
+      next unless x.marking_completed?
+      x.submissions.each do |s|
```
This one is interesting - use `current_submission_used` here.

david-yz-liu commented on app/models/admin.rb:
```diff
@@ -7,4 +7,36 @@ class Admin < User
+
+  def get_num_annotations(assignment)
+    n = 0
+    assignment.groupings.each do |x|
```
You should be able to use the [sum](http://apidock.com/rails/Enumerable/sum) method here.

david-yz-liu commented on app/models/ta.rb:
```diff
@@ -6,9 +6,9 @@ class Ta < User
   CSV_UPLOAD_ORDER = USER_TA_CSV_UPLOAD_ORDER
   SESSION_TIMEOUT = USER_TA_SESSION_TIMEOUT
 
-  after_create  :grant_repository_permissions
```
Revert these changes.

david-yz-liu commented on app/models/ta.rb:
```diff
@@ -29,6 +29,28 @@ def get_num_marked(assignment)
     n
   end
 
+  def get_num_annotations(assignment)
```
Okay, so the three methods above - make an optional parameter by a ta_id, which filters for a specific TA.

david-yz-liu commented on app/models/ta.rb:
```diff
@@ -50,24 +72,48 @@ def get_criterion_associations_by_assignment(assignment)
 
   def get_criterion_associations_count_by_assignment(assignment)
     assignment.criterion_ta_associations
-              .where(ta_id: id)
```
Revert all of these changes. You shouldn't be changing style at all.

david-yz-liu commented on app/models/ta.rb:
```diff
@@ -50,24 +72,48 @@ def get_criterion_associations_by_assignment(assignment)
+  def update_distribution(distribution, result, out_of, intervals)
+    steps = 100 / intervals # number of percentage steps in each interval
+    percentage = (result.total_mark / out_of * 100).ceil
+    if percentage == 0 then distribution[0] += 1
```
Put bodies on different lines than the conditions.

finnergizer replied:
I was getting hound errors for the method being > 10 lines so I compacted the one-line bodies. That's also why I split up the `grade_distribution_array` method into two methods as well.

david-yz-liu commented on app/views/assignments/_assignment_summary.html.erb:
```diff
@@ -14,13 +14,19 @@
+    <div id='assignment_<%= assignment.id %>_ta_graph' class='left'>
+      <%= render partial: 'graders/ta_grade_distribution_graph', locals: { assignment: assignment, ta: ta } %>
+    </div>
+    <div class="right">
```
Use single-quotes

david-yz-liu commented on app/views/assignments/_grader_summary.html.erb:
```diff
@@ -1,16 +1,39 @@
-<p class="grader_distribution"><%= I18n.t(:grader_distribution ) %></p>
+<style>
```
This should be in the CSS, not here.

david-yz-liu commented on app/views/assignments/_grader_summary.html.erb:
```diff
@@ -1,16 +1,39 @@
+</style>
 
 <% if assignment.tas.empty? %>
+  <p class="grader_distribution"><%= I18n.t(:grader_distribution) %></p>
```
Inside the views, omit `I18n`. Also, use single-quotes.

david-yz-liu commented on app/views/assignments/_grader_summary.html.erb:
```diff
@@ -1,16 +1,39 @@
+    (<%= link_to t('graders.all_graders'), { :controller => "main", :action => "grader_summary",
+                                             :assignment => assignment.id }
+  %>)</p>
+  <div style="height:380px; overflow:scroll">
```
Put this in the CSS.

david-yz-liu commented on app/views/assignments/_grader_summary.html.erb:
```diff
@@ -1,16 +1,39 @@
-    <% end %>
-  </ul>
+  <p class="grader_distribution"><%= I18n.t(:grader_distribution) %>
+    (<%= link_to t('graders.all_graders'), { :controller => "main", :action => "grader_summary",
```
Use the new style of hashes, e.g. `controller: 'main'` Also use single-quotes.

david-yz-liu commented on app/views/assignments/_list_manage.html.erb:
```diff
@@ -7,77 +7,105 @@
-            <th><%= t(:due) %></th>
-            <th><%= t(:submission_status) %></th>
-          </tr>
+        <tr>
```
I'm not going to review this part until you change the indentation back to how it was before.

david-yz-liu commented on app/views/assignments/view_summary.js.erb:
```diff
@@ -1,11 +1,15 @@
 
+jQuery('#assignment_<%= @assignment.id %>_ta_graph').html(
+    "<%= escape_javascript(render partial: 'graders/ta_grade_distribution_graph',
+                                locals: { assignment: @assignment, ta: @current_ta }) %>");
```
The indentation is off here.

david-yz-liu commented on app/views/graders/_ta_grade_distribution_graph.html.erb:
```diff
@@ -0,0 +1,35 @@
+  <h3><%= ta.first_name + " " + ta.last_name%></h3>
+  <div>
+    (<%= t('browse_submissions.how_many_marked',
+           num_marked: ta.get_num_marked(@assignment),
```
You seem to be passing in a local variable `assignment` here, which is good - you should not use the `@assignment` in this case.

david-yz-liu commented on app/views/graders/_ta_grade_distribution_graph.html.erb:
```diff
@@ -0,0 +1,35 @@
+      scaleLineWidth: 2,
+      barStrokeWidth: 1,
+      barValueSpacing: 2,
+      scaleLabel: "   <%%=value%>"
```
I'm not really sure what's going on here. Is only one `%` necessary?

david-yz-liu commented on app/views/main/grader_summary.html.erb:
```diff
@@ -0,0 +1,26 @@
+<% content_for :head do %>
```
This file should be under `assignments` not `main`

david-yz-liu commented on app/views/main/grader_summary.html.erb:
```diff
@@ -0,0 +1,26 @@
+<div class='title_bar'>
+  <h1>
+    <%= t(:assignment_link,
+            short_id: @assignment.short_identifier,
```
The indentation here is off.

david-yz-liu commented on app/views/main/grader_summary.html.erb:
```diff
@@ -0,0 +1,26 @@
+    <%= t(:assignment_link,
+            short_id: @assignment.short_identifier,
+            description: @assignment.description) %> -
+    <%=t(:grader_distribution)%>
```
There should be spaces here.

david-yz-liu commented on app/views/main/grader_summary.html.erb:
```diff
@@ -0,0 +1,26 @@
+    <div style="text-align: center">
+      <% @assignment.tas.each do |ta| %>
+        <div style="display:inline-block;">
+          <div id='assignment_<%= @assignment.id %>_ta_
```
This `id` shouldn't be split onto multiple lines

david-yz-liu commented on app/views/main/grader_summary.html.erb:
```diff
@@ -0,0 +1,26 @@
+        <div style="display:inline-block;">
+          <div id='assignment_<%= @assignment.id %>_ta_
+            <% ta.id %>_graph'>
+            <%= render partial: 'graders/ta_grade_distribution_graph', locals: { assignment: @assignment, ta: ta, a:1 } %>
```
On the other hand, this line is too long. I'm also not sure what `a` means.

finnergizer commented on app/views/graders/_ta_grade_distribution_graph.html.erb:
```diff
@@ -0,0 +1,35 @@
+      scaleLineWidth: 2,
+      barStrokeWidth: 1,
+      barValueSpacing: 2,
+      scaleLabel: '   <%%=value%>'
```
This is an option that is passed that prevents the labels on the Y-axis from appearing slightly cut off. I needed the additional % sign for it to work.

david-yz-liu commented on app/assets/stylesheets/main.scss:
```diff
@@ -82,6 +82,23 @@
+  }
+
+  ::-webkit-scrollbar {
+    -webkit-appearance: none;
```
Yeah I don't know about this one. If @arkon knows what the correct solution is great, if not I'm fine ignoring it.

arkon replied:
This is fine since it is indeed WebKit-specific. The box-shadow a few lines down should be prefix-free, however.

david-yz-liu commented on app/models/ta.rb:
```diff
@@ -15,13 +15,13 @@ class Ta < User
   end
 
-  def get_num_marked(assignment)
+  def get_num_marked(assignment, ta_id = id)
```
Make these methods only appear in the the `Assignment` model. Give it a signature `get_num_marked(assignment, ta_id = nil)` so that admins can call it without specifying an id and get everything, and tas can call it with their id to get the numbers specific to them.

david-yz-liu commented:
@finnergizer I'm quite happy with this pull request now. I've made a couple of comments, but once those are done I'm confident about merging this in.

finnergizer commented:
Thanks for the help @david-yz-liu. Made the suggested logic change and added a check for an edge case. I also refactored the assignment stat methods to be in the assignment model only, as well as removed the one vendor prefix suggested.

david-yz-liu commented on app/models/assignment.rb:
```diff
@@ -756,6 +756,55 @@ def groups_submitted
     groupings.select(&:has_submission?)
   end
 
+  def get_num_assigned(assignment, ta_id = nil)
```
This is better, but one thing I forgot to mention last time is that these are all instance methods, so passing in an extra `assignment` argument is pretty redundant.

Basically just remove all references to `assignment` in these methods. It will still work the same way.

finnergizer replied:
Yea I noticed this as well actually when I was changing it, but wasn't sure if I should stick to the signatures. I've made the changes now :).

david-yz-liu commented:
Great work, @finnergizer!
